patches/0017-mali-Avoid-splitting-lines-across-multiple-printf-calls.patch

```diff
@@ -0,0 +1,57 @@
+mali: Avoid splitting lines across multiple printf calls
+
+Since Linux 4.9, KERN_CONT is mandatory for continued lines. In the
+absence of KERN_CONT, a newline may be implicitly inserted by the core
+printk code. However, KERN_CONT should only be used by core/arch code
+during early bootup as a continued line is not SMP-safe.
+
+Avoid all these issues altogether by not splitting lines across multiple
+calls.
+
+Signed-off-by: Jonathan Liu <net147@gmail.com>
+--- a/src/devicedrv/mali/common/mali_kernel_common.h
++++ b/src/devicedrv/mali/common/mali_kernel_common.h
+@@ -113,6 +113,9 @@
+ #define MALI_PRINTF(args) _mali_osk_dbgmsg args;
+ #endif
+ 
++#define MALI_UNPAREN(...) __VA_ARGS__
++#define MALI_PREFIX_PRINTF(prefix, args) MALI_PRINTF((prefix MALI_UNPAREN args))
++
+ #define MALI_PRINT_ERROR(args) do{ \
+ 		MALI_PRINTF(("Mali: ERR: %s\n" ,__FILE__)); \
+ 		MALI_PRINTF(("           %s()%4d\n           ", __FUNCTION__, __LINE__)) ; \
+@@ -120,10 +123,7 @@
+ 		MALI_PRINTF(("\n")); \
+ 	} while(0)
+ 
+-#define MALI_PRINT(args) do{ \
+-		MALI_PRINTF(("Mali: ")); \
+-		MALI_PRINTF(args); \
```

Comment:
MALI_PRINTF seems to be defined to _mali_osk_dbgmsg, which in turn is just a "glorified" printk.

If it's just about printing a prefix, maybe we can just define pr_fmt?

Reply:
There is different prefix depending on debug level. Wouldn't that be more complex to handle with pr_fmt?

Reply:
Hmmm, indeed. As far as I can see, there's three cases: debug (MALI_DEBUG_PRINT), error (MALI_PRINT_ERROR), and "normal" (MALI_PRINT).

Maybe we can just use pr_err, pr_debug and pr_info, with pr_fmt set? That would simplify things a lot, while keeping the log levels.

Reply:
$ grep -hro 'MALI_DEBUG_PRINT([0-9]' r6p2 | sed 's/[^0-9]*//' | sort -u | xargs
1 2 3 4 5 6 7
Wouldn't we need a big switch statement in MALI_DEBUG_PRINT to call pr_alert, pr_crit, pr_err, pr_warn, pr_notice, pr_info, pr_debug according to the level argument?

The formatting of the debug level would also be changed by switching to pr_fmt.

Reply:
Sorry if that wasn't really clear, I meant ignoring the log level in MALI_DEBUG_PRINT and just forward everything to pr_debug.

Reply:
Changing MALI_PRINT_ERROR to use pr_debug would change the formatting so that all lines include the Mali prefix 
(using pr_fmt) instead of just the first line.

Reply:
Is that really an issue though? That would cause an issue only with messages that are not ended by a \n, and as you mentionned already PR_CONT is deprecated now, and grepping through it, I don't really see a lot of them.